cli/scorecard/score.go

```diff
@@ -0,0 +1,160 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scorecard
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/forseti-security/config-validator/pkg/api/validator"
+	"github.com/forseti-security/config-validator/pkg/gcv"
+	"github.com/pkg/errors"
+)
+
+// ScoringConfig holds settings for generating a score
+type ScoringConfig struct {
+	PolicyPath  string
+	categories  map[string]*constraintCategory
+	constraints map[string]*constraintViolations
+	validator   *gcv.Validator
+	ctx         context.Context
+}
+
+const otherCategoryKey = "other"
+
+// constraintCategory holds constraints by category
+type constraintCategory struct {
+	Name        string
+	constraints []*constraintViolations
+}
+
+func (c constraintCategory) Count() int {
+	sum := 0
+	for _, cv := range c.constraints {
+		sum += cv.Count()
+	}
+	return sum
+}
+
+// constraintViolations holds violations for a particular constraint
+type constraintViolations struct {
+	constraint *validator.Constraint
+	Violations []*validator.Violation `protobuf:"bytes,1,rep,name=violations,proto3" json:"violations,omitempty"`
+}
+
+func (cv constraintViolations) Count() int {
+	return len(cv.Violations)
+}
+
+func (cv constraintViolations) GetName() string {
+	return cv.constraint.GetMetadata().GetStructValue().GetFields()["name"].GetStringValue()
+}
+
+var availableCategories = map[string]string{
+	"operational-efficiency": "Operational Efficiency",
+	"security":               "Security",
+	"reliability":            "Reliability",
+	otherCategoryKey:         "Other",
+}
+
+func getConstraintForViolation(config *ScoringConfig, violation *validator.Violation) (*constraintViolations, error) {
+	key := violation.GetConstraint()
+	cv, found := config.constraints[key]
+	if !found {
+		constraint := violation.GetConstraintConfig()
+		cv = &constraintViolations{
+			constraint: constraint,
+		}
+		config.constraints[key] = cv
+
+		metadata := constraint.GetMetadata()
+		annotations := metadata.GetStructValue().GetFields()["annotations"].GetStructValue().GetFields()
+
+		categoryKey := otherCategoryKey
+		categoryValue, found := annotations["bundles.validator.forsetisecurity.org/scorecard-v1"]
+		if found {
+			categoryKey = categoryValue.GetStringValue()
+		}
+
+		category, found := config.categories[categoryKey]
+		if !found {
+			return nil, fmt.Errorf("Unknown constraint category %v for constraint %v", categoryKey, key)
+		}
+		category.constraints = append(category.constraints, cv)
+	}
+	return cv, nil
+}
+
+// attachViolations puts violations into their appropriate categories
+func attachViolations(audit *validator.AuditResponse, config *ScoringConfig) error {
```

Comment:
This and getConstraintForViolation should probably be members of ScoringConfig since they are mutating the internal state of non-exported fields.